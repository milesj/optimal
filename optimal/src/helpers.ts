import {
	Constructor,
	DefaultValue,
	DefaultValueInitializer,
	Schema,
	SchemaValidateOptions,
	UnknownObject,
} from './types';
import { ValidationError } from './ValidationError';

export function isObject(value: unknown): value is object {
	return !!value && typeof value === 'object' && !Array.isArray(value);
}

export function isSchema<T>(value: unknown): value is Schema<T> {
	return (
		isObject(value) &&
		typeof (value as UnknownObject).schema === 'function' &&
		typeof (value as UnknownObject).state === 'function' &&
		typeof (value as UnknownObject).type === 'function' &&
		typeof (value as UnknownObject).validate === 'function'
	);
}

export function isValidDate(value: Date): value is Date {
	return (
		Object.prototype.toString.call(value) === '[object Date]' &&
		!Number.isNaN(value.getTime()) &&
		value.toString() !== 'Invalid Date'
	);
}

export function isValidNumber(value: unknown): value is number {
	return typeof value === 'number' && !Number.isNaN(value);
}

export function isValidString(value: unknown): value is string {
	return typeof value === 'string' && value !== '';
}

export function createArray(value: unknown): unknown[] {
	if (value === undefined) {
		return [];
	}

	return Array.isArray(value) ? [...(value as unknown[])] : [value];
}

export function createDate(value: unknown): Date {
	if (value instanceof Date) {
		return value;
	}

	if (value === undefined || value === null) {
		return new Date();
	}

	return new Date(value as number);
}

export function createObject<T = UnknownObject>(value: unknown) {
	return (isObject(value) ? { ...value } : {}) as unknown as T;
}

/**
 * Native `instanceof` checks are problematic, as cross realm checks fail.
 * They will also fail when comparing against source and compiled files.
 * So emulate an `instanceof` check by comparing constructor names.
 */
export function instanceOf<T = unknown>(object: unknown, contract: Constructor<T>): object is T {
	if (!object || typeof object !== 'object') {
		return false;
	}

	if (object instanceof contract) {
		return true;
	}

	let current = object;

	while (current) {
		if (current.constructor.name === 'Object') {
			return false;
		}

		if (
			current.constructor.name === contract.name ||
			(current instanceof Error && current.name === contract.name)
		) {
			return true;
		}

		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
		current = Object.getPrototypeOf(current);
	}

	return false;
}

export function invalid(
	condition: boolean,
	message: string,
	path: string = '',
	value: unknown = undefined,
) {
	if (condition) {
		return;
	}

	throw new ValidationError(message, path, value);
}

export function invariant(condition: boolean, message: string) {
	if (condition) {
		return;
	}

	throw new Error(message);
}

export function logUnknown(unknownFields: object, pathPrefix?: string) {
	const unknownKeys = Object.keys(unknownFields);

	if (unknownKeys.length > 0) {
		throw new Error(
			`${pathPrefix ? `Unknown "${pathPrefix}" fields` : 'Unknown fields'}: ${unknownKeys.join(
				', ',
			)}.`,
		);
	}
}

export function pathKey(path: string): string {
	if (path.endsWith(']')) {
		const index = path.lastIndexOf('[');

		return index > 0 ? path.slice(index) : path;
	}

	const index = path.lastIndexOf('.');

	return index > 0 ? path.slice(index + 1) : path;
}

export function prettyValue(value: unknown): string | null {
	switch (typeof value) {
		case 'string':
			return `"${value}"`;

		case 'number':
		case 'function':
			return String(value);

		case 'object': {
			if (value === null) {
				return `\`null\``;
			}

			if (value.constructor !== Object) {
				return value.constructor.name === 'Array' ? null : `\`${value.constructor.name}\``;
			}

			return null;
		}

		default:
			return `\`${value}\``;
	}
}

export function tryAndCollect(
	validator: () => boolean | void,
	validError: ValidationError,
	collectErrors?: boolean,
): boolean {
	let result = false;

	try {
		const value = validator();

		if (typeof value === 'boolean') {
			result = value;
		}
	} catch (error: unknown) {
		if (error instanceof Error && collectErrors) {
			validError.addError(error);
		} else {
			throw error;
		}
	}

	return result;
}

export function extractDefaultValue<T>(
	defaultValue: DefaultValue<T>,
	path: string,
	{ currentObject, rootObject }: SchemaValidateOptions,
) {
	return typeof defaultValue === 'function'
		? (defaultValue as DefaultValueInitializer<T>)(path, currentObject!, rootObject!)
		: defaultValue;
}
