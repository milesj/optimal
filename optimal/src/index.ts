/**
 * @copyright   2021, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import * as schemas from './schemas';

export type Schemas = typeof schemas;

export { schemas };
export * from './createPredicate';
export * from './createSchema';
export * from './criteria';
export * from './optimal';
export * from './OptimalError';
export * from './schemas/array';
export * from './schemas/blueprint';
export * from './schemas/bool';
export * from './schemas/custom';
export * from './schemas/date';
export * from './schemas/func';
export * from './schemas/instance';
export * from './schemas/lazy';
export * from './schemas/number';
export * from './schemas/object';
export * from './schemas/regex';
export * from './schemas/schema';
export * from './schemas/shape';
export * from './schemas/string';
export * from './schemas/tuple';
export * from './schemas/union';
export * from './types';
export * from './ValidationError';
