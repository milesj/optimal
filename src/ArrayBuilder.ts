import Builder from './Builder';
import CollectionBuilder from './CollectionBuilder';
import { ArrayOf, DefaultValue } from './types';

export default class ArrayBuilder<T> extends CollectionBuilder<ArrayOf<T>> {
  protected contents: Builder<T> | null = null;

  constructor(contents: Builder<T> | null = null, defaultValue: DefaultValue<ArrayOf<T>> = []) {
    super('array', defaultValue);

    this.contents = contents;

    if (contents instanceof Builder) {
      this.addCheck((path, value) => {
        const nextValue = [...value];

        value.forEach((item: T, i: number) => {
          nextValue[i] = contents.run(item, `${path}[${i}]`, this.schema!)!;
        });

        return nextValue;
      });
    } else if (__DEV__ && contents) {
      this.invariant(false, 'A blueprint is required for array contents.');
    }
  }

  cast(value: unknown): ArrayOf<T> {
    return Array.isArray(value) ? value : [value];
  }

  notEmpty(): this {
    if (__DEV__) {
      this.addCheck((path, value) => {
        this.invariant(value.length > 0, 'Array cannot be empty.', path);
      });
    }

    return this;
  }

  /**
   * If contents are defined, return the type name using generics syntax.
   */
  typeAlias(): string {
    const { contents } = this;
    const alias = super.typeAlias();

    return contents ? `${alias}<${contents.typeAlias()}>` : alias;
  }
}

export function array<T = unknown>(
  contents: Builder<T> | null = null,
  defaultValue?: DefaultValue<ArrayOf<T>>,
) /* infer */ {
  return new ArrayBuilder<T>(contents, defaultValue);
}
