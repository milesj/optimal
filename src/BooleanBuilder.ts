import Builder from './Builder';
import { DefaultValue } from './types';

export default class BooleanBuilder<T extends boolean = boolean> extends Builder<T> {
  constructor(defaultValue?: DefaultValue<T>) {
    super('boolean', defaultValue || (false as T));
  }

  onlyFalse(): BooleanBuilder<false> {
    this.defaultValue = false as T;

    if (__DEV__) {
      this.addCheck((path, value) => {
        this.invariant(value === false, 'May only be `false`.', path);
      });
    }

    return (this as unknown) as BooleanBuilder<false>;
  }

  onlyTrue(): BooleanBuilder<true> {
    this.defaultValue = true as T;

    if (__DEV__) {
      this.addCheck((path, value) => {
        this.invariant(value === true, 'May only be `true`.', path);
      });
    }

    return (this as unknown) as BooleanBuilder<true>;
  }
}

export function bool(defaultValue: DefaultValue<boolean> = false) /* infer */ {
  return new BooleanBuilder<boolean>(defaultValue);
}
