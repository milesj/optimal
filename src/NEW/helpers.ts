import { Constructor, UnknownObject, Schema } from './types';

export function createDate(value: unknown): Date {
  return value instanceof Date ? value : new Date(value as number);
}

/**
 * Native `instanceof` checks are problematic, as cross realm checks fail.
 * They will also fail when comparing against source and compiled files.
 * So emulate an `instanceof` check by comparing constructor names.
 */
export function instanceOf<T = unknown>(
  object: unknown,
  contract: Constructor<T> | Function,
): object is T {
  if (!object || typeof object !== 'object') {
    return false;
  }

  if (object instanceof contract) {
    return true;
  }

  let current = object;

  while (current) {
    if (current.constructor.name === 'Object') {
      return false;
    }

    if (
      current.constructor.name === contract.name ||
      (current instanceof Error && current.name === contract.name)
    ) {
      return true;
    }

    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    current = Object.getPrototypeOf(current);
  }

  return false;
}

export function invariant(condition: boolean, message: string, path: string = '') {
  if (__DEV__) {
    if (condition) {
      return;
    }

    // TODO no errorPrefix
    const prefix = path ? `Invalid field "${path}".` : '';

    throw new Error(`${prefix} ${message}`.trim());
  }
}

export function isObject(value: unknown): value is object {
  return !!value && typeof value === 'object' && !Array.isArray(value);
}

export function isSchema<T>(value: unknown): value is Schema<T> {
  return (
    isObject(value) &&
    typeof (value as UnknownObject).validate === 'function' &&
    typeof (value as UnknownObject).type === 'function'
  );
}

export function isValidDate(value: Date): value is Date {
  return (
    Object.prototype.toString.call(value) === '[object Date]' &&
    !Number.isNaN(value.getTime()) &&
    value.toString() !== 'Invalid Date'
  );
}

export function isValidNumber(value: unknown): value is number {
  return typeof value === 'number' && !Number.isNaN(value);
}

export function isValidString(value: unknown): value is string {
  return typeof value === 'string' && value !== '';
}

export function pathKey(path: string): string {
  const index = path.lastIndexOf('.');

  return index > 0 ? path.slice(index + 1) : path;
}
