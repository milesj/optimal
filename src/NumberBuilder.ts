import Builder from './Builder';
import { DefaultValue } from './types';

function isNumber(value: unknown): value is number {
  return typeof value === 'number';
}

export default class NumberBuilder<T extends number = number> extends Builder<T> {
  constructor(defaultValue?: DefaultValue<T>) {
    super('number', defaultValue || (0 as T));
  }

  between(min: number, max: number, inclusive: boolean = false): this {
    if (__DEV__) {
      this.invariant(
        isNumber(min) && isNumber(max),
        'Between requires a minimum and maximum number.',
      );

      this.addCheck((path, value) => {
        this.invariant(
          isNumber(value) &&
            (inclusive ? value >= min && value <= max : value > min && value < max),
          `Number must be between ${min} and ${max}${inclusive ? ' inclusive' : ''}.`,
          path,
        );
      });
    }

    return this;
  }

  gt(min: number, inclusive: boolean = false): this {
    if (__DEV__) {
      this.invariant(isNumber(min), 'Greater-than requires a minimum number.');

      this.addCheck((path, value) => {
        if (inclusive) {
          this.invariant(
            isNumber(value) && value >= min,
            `Number must be greater than or equal to ${min}.`,
            path,
          );
        } else {
          this.invariant(
            isNumber(value) && value > min,
            `Number must be greater than ${min}.`,
            path,
          );
        }
      });
    }

    return this;
  }

  gte(min: number): this {
    return this.gt(min, true);
  }

  lt(max: number, inclusive: boolean = false): this {
    if (__DEV__) {
      this.invariant(isNumber(max), 'Less-than requires a maximum number.');

      this.addCheck((path, value) => {
        if (inclusive) {
          this.invariant(
            isNumber(value) && value <= max,
            `Number must be less than or equal to ${max}.`,
            path,
          );
        } else {
          this.invariant(isNumber(value) && value < max, `Number must be less than ${max}.`, path);
        }
      });
    }

    return this;
  }

  lte(max: number): this {
    return this.lt(max, true);
  }

  oneOf<U extends number>(list: U[]): NumberBuilder<U> {
    if (__DEV__) {
      this.invariant(
        Array.isArray(list) && list.length > 0 && list.every(item => isNumber(item)),
        'One of requires a non-empty array of numbers.',
      );

      this.addCheck((path, value) => {
        this.invariant(list.includes(value), `Number must be one of: ${list.join(', ')}`, path);
      });
    }

    return (this as unknown) as NumberBuilder<U>;
  }
}

export function number<T extends number = number>(defaultValue?: DefaultValue<number>) /* infer */ {
  return new NumberBuilder<T>(defaultValue as T);
}
