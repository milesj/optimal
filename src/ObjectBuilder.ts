import Builder from './Builder';
import CollectionBuilder from './CollectionBuilder';
import { ObjectOf, DefaultValue } from './types';
import { builder } from './InstanceBuilder';
import isObject from './isObject';

export default class ObjectBuilder<T> extends CollectionBuilder<ObjectOf<T>> {
  protected contents: Builder<T> | null = null;

  constructor(contents: Builder<T> | null = null, defaultValue: DefaultValue<ObjectOf<T>> = {}) {
    super('object', defaultValue);

    this.contents = contents;

    if (contents instanceof Builder) {
      this.addCheck((path, value) => {
        const nextValue = { ...value };

        Object.keys(value).forEach(key => {
          nextValue[key] = contents.run(value[key], `${path}.${key}`, this.schema!)!;
        });

        return nextValue;
      });
    } else if (__DEV__ && contents) {
      this.invariant(false, 'A blueprint is required for object contents.');
    }
  }

  cast(value: unknown): ObjectOf<T> {
    const obj = isObject(value) ? value : {};

    return (obj as unknown) as ObjectOf<T>;
  }

  notEmpty(): this {
    if (__DEV__) {
      this.addCheck((path, value) => {
        this.invariant(Object.keys(value).length > 0, 'Object cannot be empty.', path);
      });
    }

    return this;
  }

  /**
   * If contents are defined, return the type name using generics syntax.
   */
  typeAlias(): string {
    const { contents } = this;
    const alias = super.typeAlias();

    return contents ? `${alias}<${contents.typeAlias()}>` : alias;
  }
}

export function object<T = unknown>(
  contents: Builder<T> | null = null,
  defaultValue?: DefaultValue<ObjectOf<T>>,
) /* infer */ {
  return new ObjectBuilder<T>(contents, defaultValue);
}

export function blueprint<T = unknown>(
  defaultValue?: DefaultValue<ObjectOf<Builder<T>>>,
) /* infer */ {
  return new ObjectBuilder<Builder<T>>(builder<T>().notNullable(), defaultValue);
}
