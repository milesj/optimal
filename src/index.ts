/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import optimal from './optimal';
import Builder, { custom, func } from './Builder';
import ArrayBuilder, { array } from './ArrayBuilder';
import BooleanBuilder, { bool } from './BooleanBuilder';
import InstanceBuilder, { builder, instance, date, regex } from './InstanceBuilder';
import NumberBuilder, { number } from './NumberBuilder';
import ObjectBuilder, { object, blueprint } from './ObjectBuilder';
import ShapeBuilder, { shape } from './ShapeBuilder';
import StringBuilder, { string } from './StringBuilder';
import TupleBuilder, { tuple } from './TupleBuilder';
import UnionBuilder, { union } from './UnionBuilder';

export const predicates = {
  array,
  blueprint,
  bool,
  builder,
  custom,
  date,
  func,
  instance,
  number,
  object,
  regex,
  shape,
  string,
  tuple,
  union,
};

export {
  array,
  blueprint,
  bool,
  builder,
  custom,
  date,
  func,
  instance,
  number,
  object,
  regex,
  shape,
  string,
  tuple,
  union,
};

export {
  Builder,
  ArrayBuilder,
  BooleanBuilder,
  InstanceBuilder,
  NumberBuilder,
  ObjectBuilder,
  ShapeBuilder,
  StringBuilder,
  TupleBuilder,
  UnionBuilder,
};

export type Predicates = typeof predicates;

export * from './types';

export default optimal;
