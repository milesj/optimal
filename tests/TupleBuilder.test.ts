import TupleBuilder, { tuple } from '../src/TupleBuilder';
import { array } from '../src/ArrayBuilder';
import { bool } from '../src/BooleanBuilder';
import { object } from '../src/ObjectBuilder';
import { number } from '../src/NumberBuilder';
import { string } from '../src/StringBuilder';
import { runChecks } from './helpers';

describe('TupleBuilder', () => {
  type Tuple = [string[], boolean, number, object, string];
  let builder: TupleBuilder<Tuple>;

  beforeEach(() => {
    builder = tuple([
      array(string()),
      bool(true),
      number(1).between(0, 5),
      object(number()),
      string('foo').oneOf(['foo', 'bar', 'baz']),
    ]);
  });

  it('errors if a non-array is not passed', () => {
    expect(() => {
      // @ts-ignore
      union('foo', []);
    }).toThrowErrorMatchingSnapshot();
  });

  it('errors if an empty array is passed', () => {
    expect(() => {
      tuple(
        // @ts-ignore
        [],
      );
    }).toThrowErrorMatchingSnapshot();
  });

  it('errors if an array with non-builders is passed', () => {
    expect(() => {
      tuple([
        // @ts-ignore
        123,
      ]);
    }).toThrowErrorMatchingSnapshot();
  });

  it('doesnt error if a builder array is passed', () => {
    expect(() => {
      tuple<[string]>([string()]);
    }).not.toThrow();
  });

  describe('notEmpty()', () => {
    beforeEach(() => {
      builder.notEmpty();
    });

    it('errors if value is empty', () => {
      expect(() => {
        runChecks(builder, []);
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if not empty', () => {
      expect(() => {
        runChecks(builder, [['123']]);
      }).not.toThrow();
    });
  });

  describe('typeAlias()', () => {
    it('returns the type name', () => {
      expect(builder.typeAlias()).toBe('[array<string>, boolean, number, object<number>, string]');
    });
  });
});
