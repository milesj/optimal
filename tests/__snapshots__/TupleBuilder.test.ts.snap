// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TupleBuilder errors if a non-array is not passed 1`] = `"union is not defined"`;

exports[`TupleBuilder errors if an array with non-builders is passed 1`] = `"A non-empty array of blueprints are required for a tuple."`;

exports[`TupleBuilder errors if an empty array is passed 1`] = `"A non-empty array of blueprints are required for a tuple."`;

exports[`TupleBuilder notEmpty() errors if value is empty 1`] = `"Invalid field \\"key[2]\\". Number must be between 0 and 5."`;
